### ISSUE TITLE
Less helpful compiler error upon module/trait name clashes

### ISSUE THREAD

**Issue by tommythorn** [rust-lang/rust#85345]
I was encouraged [1] to report this and filing as a bug seemed the most appropriate.  The issue is of course clear to the experienced user, but it was very confusing.  The issue is that moving a trait into a module with the same name leads to confusing compiler errors.  It's possible this could be improved by noticing the clash which is bound to lead to problems.  The full repro [2] has not been boiled down as it's simple enough and I wanted to give the context to motivate why I would have done this, however the gist is this:

[1] https://twitter.com/ekuber/status/1393605930483388416
[2] https://github.com/tommythorn/bp/tree/newb-struggles

```rust

```

I expected to see this happen: *explanation*

Instead, this happened: *explanation*

### Meta

`rustc --version --verbose`:
```
mod Boolish { // XXX
    pub trait Boolish {
        fn update(&mut self, taken: bool) -> &mut Self;
        fn value(&self) -> bool;
        fn new(b: bool) -> Self;
    }
...
    impl Boolish for TwoBitCounter {
        fn update(&mut self, taken: bool) -> &mut Self {
            ...
        }
...
} // XXX
use Boolish::*; // XXX
...
    fn new(addr_bits: usize) -> LocalBp {
        let pht = vec![TwoBitCounter::new(true); 1 << addr_bits];
```

Without the lines marked with XXXs this works perfectly.

**Comment by tommythorn** [comment]
Actually, I still don't understand the error even knowing the collision.  It makes no sense to me.

**Comment by estebank** [comment]
Did you try [following the suggestion](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=961036be061225f661f1d29df2852577)?

```
  help: the following trait is implemented but not in scope; perhaps add a `use` for it:
      |
  2   | use crate::Boolish::Boolish;
      |
```

[That should have given you an error like](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=fb4643750cfc3d1092077dd217b73351):

```
error[E0255]: the name `Foo` is defined multiple times
  --> f21.rs:12:5
   |
1  | mod Foo {
   | ------- previous definition of the module `Foo` here
...
12 | use Foo::Foo;
   |     ^^^^^^^^ `Foo` reimported here
   |
   = note: `Foo` must be defined only once in the type namespace of this module
help: you can use `as` to change the binding name of the import
   |
12 | use Foo::Foo as OtherFoo;
   |     ^^^^^^^^^^^^^^^^^^^^
```

After [following _that_ suggestion, the code would have worked](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=1f0b7dd66634ea20864bcbff8a297fab).

----

This happens because we explicitly avoid clobbering locally available names in the type namespace with an `use foo::*;` when that name is already present.

----

There are a few things we could do:

 - not show the span to the ADT's definition when we find a trait that would apply, that just adds noise
 - actively detect cases of shadowing and provide a specific suggestion, particularly to avoid the second error
 - if a blanket `use` is at play, try to explain why the explicit `use` is also needed

**Comment by tommythorn** [comment]
I  did, although first naively adding the `use crate Boolish::Boolish;` under the original `use Boolish::*;` leads to a different complaint.  Remove the original `use Boolish::*;` did get me to the message you described which I guess I should have meditated on a bit longer (my only excuse is that between the three variations,  it wasn't obvious that I was actually making progress).

I think the reason it didn't click is because the message above "previous definition of the module `Foo`" *only* refers to "module `Foo`" so I missed that it was a conflict between the trait name and the module name (it read as if it was a conflict between two modules).

I'm not sure what would have been better, but the second item sounds like it directly addresses my mistake.

Thanks.
